### PR TITLE
Rachana - Bumblebees

### PR DESCRIPTION
**Pull request by rachanapatel**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,14 +1,41 @@
...
+      }
+    }
+    return count;
+  }
+
+
   return (
     <div id="App">
       <header>
         <h1>Application title</h1>
```
Comment (anselrognlie) [inline src/App.jsx:34]: It would be nice to update this to at least a hard-coded representation of the conversation title: 
```
<h1>Chat between Vladimir and Estragon</h1>
```

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,32 @@
...
   return (
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp}></TimeStamp></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:15]: Nice use of the supplied `TimeStamp`. We pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Note that for components that don't expect to be given child markup, rather than explicitly writing the closing tag `<TimeStamp></TimeStamp>`, we prefer to use the empty tag shorthand `<TimeStamp />`.

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onHeart: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:29]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+ChatLog.propTypes = {
+    entries: PropTypes.arrayOf(PropTypes.shape({
+        id: PropTypes.number.isRequired,
+        sender: PropTypes.string.isRequired,
+        body: PropTypes.string.isRequired,
+        timeStamp: PropTypes.string.isRequired,
+        liked: PropTypes.bool.isRequired,
+    })).isRequired,
+    onHeart: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:21]: Similar to the props for `ChatEntry`, here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = (props) => {
+    return (
+        props.entries.map(entry => 
+        <ChatEntry key={entry.id} id={entry.id} sender={entry.sender} body={entry.body} liked={entry.liked} timeStamp= {entry.timeStamp}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:8]: We should still aim for lines about 80 columns wide, even when not governed by a recommendation like Python's PEP8. And when our components don't accept nested components (when the tag content body is empty), prefer to omit the closing tag and use the self closing tag syntax (e.g., `<MyComponent/>` rather than `<MyComponent></MyComponent>`).
```
        <ChatEntry
            key={entry.id}
            id={entry.id}
            sender={entry.sender}
            body={entry.body}
            liked={entry.liked}
            timeStamp={entry.timeStamp}
            onHeart={props.onHeart}
        />
```

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = (props) => {
+    return (
+        props.entries.map(entry =>
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:7]: While not required in this case (since you are directly returning an array of components, each with a `key`), consider including a wrapper `div` around the list of `ChatEntry` components, as this provides a convenient place to add a `className` for the list of messages for styling purposes.

File: src/components/ChatEntry.jsx [hunk 7]
```diff
@@ -1,20 +1,32 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+import { useState } from 'react';
+
+const ChatEntry = (props) => {
+  // const [heart, setHeart] = useState(props.liked);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 8]
```diff
@@ -1,20 +1,32 @@
...
+  let heartShown = props.liked ? '❤️' : '🤍';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:8]: 👍 Nice use of a local variable to store the emoji picked using a ternary. Consider moving even something small like this to a helper function external to the component, so that the behavior of the emoji-picking logic can be tested easily outside of React.

File: src/components/ChatEntry.jsx [hunk 9]
```diff
@@ -1,20 +1,32 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp}></TimeStamp></p>
+        {/* <button className="like" onClick={() => setHeart(!heart)}>{heartShown}</button> */}
+        <button className="like" onClick={() => props.onHeart(props.id)}>{heartShown}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:17]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,41 @@
...
+  const [messages, setMessages] = useState(jsonData);
+
+  const updateMsgLikes = (id) => {
+    const update = messages.map(message => {
+          if (message.id === id) {
+            return { ...message, liked: !message.liked};
+          }
+          return message;
+        });
+        return setMessages(update)
```
Comment (anselrognlie) [inline src/App.jsx:17]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setMessages(messages => messages.map(message => {
      if (message.id === id) {
        return {...message, liked: !message.liked};
      } else {
        return message;
      };
    }));
```
We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,41 @@
...
+
+  const numHearts = () => {
+    let count = 0;
+    for (const msg of messages) {
+      if (msg.liked === true) {
+        count++;
+      }
+    }
+    return count;
+  }
```
Comment (anselrognlie) [inline src/App.jsx:28]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Explicitly totalling the count is perfectly fine, but many JS programmers would use `reduce` to achieve this.

Also, notice that we could move this helper entirely out of the component if we passed the current `messages` data as a parameter, rather than using enclosing scope to access it.
```js
  const numHearts = (messages) => {
    return messages.reduce((count, msg) => {
      return msg.liked ? count + 1 : count;
    }, 0);
  };
```
The first few times we work with `reduce`, it can be challenging to understand, but it's a tool that gets used commonly enough that it's worth practicing.

File: src/components/ChatEntry.jsx [hunk 12]
```diff
@@ -1,20 +1,32 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+import { useState } from 'react';
+
+const ChatEntry = (props) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: Personally, I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. Even though we do need to remember that these are all passed in as a single object (your `props` parameter) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params), I really prefer the glanceability.

If you changed your approach to use destructuring here, you'd most likely want to be consistent across all your components.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
